src/dropout.jl

```diff
@@ -0,0 +1,156 @@
+
+"""
+    dropout([rng], A, p; dims=:)
+
+Returns an array in which each element of `A` is either replaced with zero,
+with probability `p`, or else multiplied by `1/(1-p)`.
+
+By default every element is treated independently.
+With `dims=1`, a choice is made for every value of the 1st index
+i.e. each row of a matrix is either zero or not.
+
+Optional first argument is the random number generator used.
+
+# Examples
+```
+julia> dropout(ones(2, 10), 0.2)
+2×10 Matrix{Float64}:
+ 1.25  1.25  0.0   1.25  1.25  1.25  1.25  1.25  1.25  1.25
+ 1.25  1.25  1.25  0.0   1.25  1.25  0.0   1.25  1.25  1.25
+
+julia> mean(dropout(ones(10^4, 5), 0.2), dims=1)
+1×5 Matrix{Float64}:
+ 0.998  1.00075  0.99125  0.99575  1.00075
+
+julia> dropout(ones(5, 5), 0.7, dims=1)  # whole row the same
+5×5 Matrix{Float64}:
+ 3.33333  3.33333  3.33333  3.33333  3.33333
+ 0.0      0.0      0.0      0.0      0.0
+ 0.0      0.0      0.0      0.0      0.0
+ 3.33333  3.33333  3.33333  3.33333  3.33333
+ 0.0      0.0      0.0      0.0      0.0
+
+julia> mean(dropout(ones(10^4, 5), 0.3, dims=1), dims=1)
+1×5 Matrix{Float64}:
+ 1.00571  1.00571  1.00571  1.00571  1.00571
+```
+"""
+dropout(A::AbstractArray, p::Real; dims = :) = dropout(_rng_from_array(A), A, p; dims)
+
+function dropout(rng::AbstractRNG, A::AbstractArray, p::Real; dims = :)
+    T = float(eltype(A))
+    0 <= p <= 1 || throw(ArgumentError("dropout expects a probability 0 <= p <= 1"))
+    if p > 0
+        dst = similar(A, T)
+        pT = convert(real(T), p)
+        _dropout!(rng, dst, A, pT, dims)
+    else
+        # Not so sure we want fast paths... this tries but doesn't guarantee type-stability,
+        # and the rrule does not have such a fast paths.
+        convert(AbstractArray{T}, A)
+    end
+end
+
+"""
+    dropout!(B, A, p; dims=:)
+
+This does exactly `B .= dropout(A, p; dims)`,
+or rather, it's the implementation of out-of-place [`dropout`](@ref).
+"""
+function dropout!(dst::AbstractArray, src::AbstractArray, p::Real; dims=:)
```

Comment:
IMO if there is a single `dropout!` function, it should have an explicit `rng` parameter. Adding an overload that uses `_rng_from_array` like `dropout` would work too.

Reply:
ok I made two methods now, optional RNG.